### ISSUE TITLE
Improve ordering of content

### ISSUE THREAD

**Issue by milesfrain** [purescript-contrib/purescript-book#340]
Some of the book's content is presented in a non-ideal sequence which frustrates beginners in the following ways:
- We present exercises that cannot be solved idiomatically using only the material covered so far - and even worse, is that some of our reference solutions use techniques that are only covered in later chapters.
- We refer to stuff that has not been explained yet. Chapter 4 is a pretty bad offender (references `Maybe` before ADTs, `Unit` with no explanation, `do` notation and `pure` without discussing `bind` and related typeclasses).

One simple improvement is to cover pattern matching (ch5) before recursion (ch4) (highlighted in #303). We should also cover ADTs (and `Maybe`) before `do` notation (proposed in #212).

But I think the book would really benefit from a more systematic restructuring where we do something like the following:
- Create a list of all topics we want the book to cover.
- Identify which topics depend on other topics.
- Create new (or classify existing) exercises for each topic and note what other prerequisites are required for writing idomatic solutions.
- Assemble these topics in the book where we're respecting all prerequisites.

We could also push more of these topic explanations to other language reference locations (e.g. [docs repo](https://github.com/purescript/documentation/) and library docs), then link to those from the book. That way, the book will serve more as a guide on learning sequence with accompanying exercises to avoid duplicating content.

I think there would also be value of distinguishing a minimum set of "core" concepts that all users _must_ know (e.g. functions, ADTs, typeclasses, etc. - noted in https://discourse.purescript.org/t/guide-to-learn-purescript/2308/4) from the supplemental topics (e.g. generative testing, canvas graphics). This could be noted in the book's introduction.

**Comment by klarkc** [comment]
I just found this case [right here](https://github.com/purescript-contrib/purescript-book/blob/5979144a9d179c83a9147dac29d329378a05846c/exercises/chapter4/no-peeking/Solutions.purs#L30). I was reviewing the solutions that were made to compare to my solutions, and did not expected a pattern matching solution that only appears on the next chapter.

**Comment by Zelenya** [comment]
I did a pass through the issues collecting all the suggestions and through code gathering the topics.

**TLDR. I don’t think we need to redo or move a lot of stuff around. I propose a few swaps (for example, swap chapters 4 and 5) and then extend content based on the existing (or future) issues and confusions.**

Additionally, I’d like to break (some) chapters into subchapters to make it easier to navigate and digest. (see [Rust book](https://doc.rust-lang.org/book/)

> Note that I’ve been focusing on content – not exercises; I think these could be added ad-hoc afterwards.

## Context

### Current topics

- **Chapter 3.** Functions and Records
    - Types
    - Quantified Types
    - Indentation
    - Records
    - Type Constructors
    - Kinds
    - Functions
    - Curried Functions
    - Infix Function Application
    - Function Composition
- **Chapter 4.** Recursion, Maps, And Fold
    - *(uses maybes, while they are introduced in chapter 5)*
    - Recursion
    - Map and filter
    - Array Comprehensions
    - Do Notation
    - Folds
    - Tail Recursion
- **Chapter 5**. Pattern Matching
    - Pattern Matching
    - Array Patterns
    - Row Polymorphism
    - Record Puns
    - Algebraic Data Types
    - Newtypes
- **Chapter 6**. Type Classes
    - Show, Eq, Ord, Field
    - Semigroups, Monoids, Foldable
    - Functor
    - Deriving Instances
    - Type Class Constraints
    - Multi Parameter Type Classes
    - Functional Dependencies
    - Superclasses
- **Chapter 7.** Applicative
    - Applicative
    - Traversable
- **Chapter 8.** Effect
    - Monad
    - Side-Effects and Purity
    - Exceptions
    - Mutable State
    - DOM Effects
- **Chapter 9.** Asynchronous Effects
    - Aff
    - Parallel Computations
- **Chapter 10.** The Foreign Function Interface
    - Asynchronous Functions
    - JSON
    - Calling PureScript from JavaScript
- **Chapter 11.** Monads
    - State
    - Reader
    - Writer
    - Monad Transformers
    - mtl
    - Alternatives
- **Chapter 12. Canvas Graphics**
- **Chapter 13. Testing**
- **Chapter 14. Domain-Specific Languages**

### Current/open issues

- [ ]  Update intro https://github.com/purescript-contrib/purescript-book/pull/193 https://github.com/purescript-contrib/purescript-book/issues/173
- [ ]  Recheck all the repl type-outputs
    - [ ]  Add a note on constraints when they first pop-up in the repl signature (type-classes come late, the type-signatures shouldn’t be super mysterious until then)
- [ ]  Drop type class instance names
    - [ ]  Chapter 6 should still have a note on naming
- [ ]  Swap **Chapter 4** and **Chapter 5**  https://github.com/purescript-contrib/purescript-book/issues/340
- [ ]  Chapter 3. Fix notes on `where` indentation https://github.com/purescript-contrib/purescript-book/issues/397
- [ ]  Chapter 3. Move `flip` to Chapter 4 with more description and new repl output https://github.com/purescript-contrib/purescript-book/issues/326 https://github.com/purescript-contrib/purescript-book/issues/260
- [ ]  Chapter 3. Improve curried functions and partial application https://github.com/purescript-contrib/purescript-book/issues/261
- [ ]  Chapter 4. `Unit` with no explanation https://github.com/purescript-contrib/purescript-book/issues/340
- [ ]  Chapter 4. `do` notation and `pure` without discussing `bind` and related typeclasses. https://github.com/purescript-contrib/purescript-book/issues/340
    - [ ]  Explain abstractions vs concrete map, pure, etc.
    - [ ]  Reference/Cover it later in the book
- [ ]  ~~Chapter 4. Exercise. countEven wants (future topic) pattern matching https://github.com/purescript-contrib/purescript-book/issues/303~~
    - [ ]  ~~Solved by chapter 4-5 swap~~
- [ ]  ~~Chapter 4. Cover ADTs (and `Maybe`) before `do` notation https://github.com/purescript-contrib/purescript-book/issues/212~~
    - [ ]  ~~Partially handled by the swap and prev. issue~~
- [ ]  Chapter 6. Improve explanation of Multi Parameter Type Classes and Functional Dependencies https://github.com/purescript-contrib/purescript-book/issues/238
- [ ]  Chapter 6. ****Add content on deriving instances https://github.com/purescript-contrib/purescript-book/issues/149
- [ ]  Chapter 7. Discuss zippy Array and List https://github.com/purescript-contrib/purescript-book/issues/236
- [ ]  Chapter 7. Align definitions of `combineList` and `sequence` https://github.com/purescript-contrib/purescript-book/issues/155
- [ ]  Chapter 8. Clarify Monads and Applicatives https://github.com/purescript-contrib/purescript-book/issues/186
- [ ]  Chapter 8. Clarify `Either e` https://github.com/purescript-contrib/purescript-book/issues/162
- [ ]  Chapter 11. Add example(s) of Reader https://github.com/purescript-contrib/purescript-book/issues/412
- [ ]  Chapter 13. Check QuickCheck ****with docs https://github.com/purescript-contrib/purescript-book/issues/188
- [ ]  New. Add a spy (Chapter 4 or somewhere else) https://github.com/purescript-contrib/purescript-book/issues/270
- [ ]  New. Consider introducing typed holes (Chapter 6 or extra) https://github.com/purescript-contrib/purescript-book/issues/233
- [ ]  New Chapter. Add content on Row Types and Variants https://github.com/purescript-contrib/purescript-book/issues/60
